thali/thalireplicationmanager.js

```diff
@@ -33,30 +43,93 @@ ThaliReplicationManager.events = {
   START_ERROR: 'startError',
   STOP_ERROR: 'stopError',
   CONNECT_ERROR: 'connectError',
-  DISCONNECT_ERROR: 'disconnectError',
-  SYNC_ERROR: 'syncRetry'
+  DISCONNECT_ERROR: 'disconnectError'
 };
 
 /**
-* Starts the Thali replication manager with the given device name and port number
-* @param {String} deviceName the device name to advertise.
+* Returns the current device's id if available. If not, the provided callback
+* function is saved and called when the id does become available.
+* @param {Function} cb the callback which returns the current device's id.
+*/
+ThaliReplicationManager.prototype.getDeviceIdentity = function (cb) {
+  if(this._deviceIdentityFlag == deviceIdentityFlag.deviceIdentityAvailable) {
+    cb(null, this._deviceName);
+    return;
+  }
+  if(this._deviceIdentityFlag == deviceIdentityFlag.noDeviceIdentitySet) {
+    this._deviceIdentityFlag = deviceIdentityFlag.gettingDeviceIdentity;
+    getPublicKeyHash.call(this, cb);
+    return;
+  }
+  if(this._deviceIdentityFlag == deviceIdentityFlag.gettingDeviceIdentity) {
+    // save the callback for future
+    this._deviceIdentityListeners.push(cb);
+  }
+}
+
+function getPublicKeyHash(cb) {
+  cryptomanager.getPublicKeyHash(function (err, publicKeyHash) {
```

Comment:
I have several suggestions here.
1. I would suggest that when getDeviceIdentity is called above and you end up in line 66 that you immediately just add cb to _deviceIdentityListeners. Otherwise you end up with the code below where you have to remember (in two different places) to call informDeviceIdentityListeners as well as to clean up the cb. If you just add cb to _deviceIdentityListeners then you won't have to remember anything.
2. Why do you set this_deviceIdentityListeners = [] on both 75 and 81? Since you have to do it in both places why not just set that line after line 83?
3. If you implement the two suggestions above then this function gets small enough that I would suggest just getting rid of it and putting the smaller code straight into line 61.